src/participant/prompts/history.ts

```diff
@@ -1,37 +1,151 @@
   const messages: vscode.LanguageModelChatMessage[] = [];
 
-  context.history.map((historyItem) => {
-    if (
-      historyItem.participant === CHAT_PARTICIPANT_ID &&
-      historyItem instanceof vscode.ChatRequestTurn
-    ) {
+  for (const historyItem of context.history) {
+    if (historyItem instanceof vscode.ChatRequestTurn) {
+      if (
+        historyItem.prompt?.trim().length === 0 ||
+        connectionNames?.includes(historyItem.prompt)
+      ) {
+        // When the message is empty or a connection name then we skip it.
+        // It's probably going to be the response to the connect step.
+        continue;
+      }
+
       // eslint-disable-next-line new-cap
       messages.push(vscode.LanguageModelChatMessage.User(historyItem.prompt));
     }
 
     if (historyItem instanceof vscode.ChatResponseTurn) {
-      let res = '';
+      let message = '';
+
+      if (
+        (historyItem.result as EmptyRequestChatResult).metadata
+          ?.isEmptyResponse ||
+        (historyItem.result as AskToConnectChatResult).metadata?.askToConnect
+      ) {
+        // Skip a response to an empty user prompt message or connect message.
+        continue;
+      }
+
       for (const fragment of historyItem.response) {
-        if (
-          fragment instanceof vscode.ChatResponseMarkdownPart &&
-          historyItem.result.metadata?.responseContent
-        ) {
-          res += fragment.value.value;
+        if (fragment instanceof vscode.ChatResponseMarkdownPart) {
+          message += fragment.value.value;
+
+          if (
+            (historyItem.result as NamespaceRequestChatResult).metadata
+              ?.askForNamespace
+          ) {
+            // When the message is the assistant asking for part of a namespace,
+            // we only want to include the question asked, not the user's
+            // database and collection names in the history item.
+            break;
+          }
         }
       }
       // eslint-disable-next-line new-cap
-      messages.push(vscode.LanguageModelChatMessage.Assistant(res));
+      messages.push(vscode.LanguageModelChatMessage.Assistant(message));
     }
-  });
+  }
 
   return messages;
 }
+
+// A user could mistakenly write one of these which would mess up our parsing.
+// To fix this we either need to pass metadata with the command, which
+// is doesn't look like it possible, or we need to have a more complex
+// message key, like using a special symbol for this regex (although a user could
+// copy paste this themselves and get into another weird state).
+export const databaseNameInCommandResponseKey = 'Database: ';
+const databaseNameInCommandResponseKeyRegex = `^${databaseNameInCommandResponseKey}(.*)`;
+export const collectionNameInCommandResponseKey = 'Collection: ';
+const collectionNameInCommandResponseKeyRegex = `^${collectionNameInCommandResponseKey}(.*)`;
+
+// eslint-disable-next-line complexity
+export function getLatestDatabaseAndCollectionFromChatHistory({
```

Comment:
Do we actually need to store the last namespace this way? Couldn't we rely on the previous conversation? We could also include in the prompt the last generated code block, so the model knows what was the last state of query generation.

Reply:
> Do we actually need to store the last namespace this way?
We don't need to. This provides a fallback if the namespace request to the model cannot find the namespace in the chat. 
> Couldn't we rely on the previous conversation?
Do you mean the chat messages that have been passed so far? Sometimes the model doesn't find it and the fallback can be helpful there.